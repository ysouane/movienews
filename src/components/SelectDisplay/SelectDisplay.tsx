import React from "react"
import "./SelectDisplay.scss"

const SelectDisplay = ({ selected, changeDisplay }: string | any) => {
  return (
    <label htmlFor="typeOfDisplay">
      <select
        className="selectDisplay"
        id="typeOfDisplay"
        onChange={changeDisplay}
        defaultValue={selected}
      >
        <option value="recentMovies">Films r√©cent</option>
        <option value="popularMovies">Films les plus populaires</option>
        <option value="favorites">Favoris</option>
      </select>
    </label>
  )
}

export default SelectDisplay
