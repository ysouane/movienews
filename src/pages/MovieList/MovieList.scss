@import 'styles/variables';

.movieList {
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: $white;

  &__selectDisplay {
    display: flex;
    justify-content: center;
    margin: 2rem;
  }

  &__card {
    margin-left: 3rem;
    margin-right: 3rem;
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex-wrap: wrap;
    gap: 2%;

    @media screen and (min-width: $breakpoints-md) {
      flex-direction: row;
    }
  }
}