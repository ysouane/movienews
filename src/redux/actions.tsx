export const changeTypeOfDisplayAndUrl = (value: string) => ({
  type: "changeTypeOfDisplayAndUrl",
  payload: { value }
})
