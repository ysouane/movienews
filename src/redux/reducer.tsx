import { API_KEY, API_URL, API_VERSION } from "helpers"
import currentDate from "utils/currentDateUTCFormatted"

export default function Reducer(state: any, action: any) {
  if (action.type === "changeTypeOfDisplayAndUrl") {
    const { value } = action.payload

    if (value === "recentMovies") {
      return {
        ...state,
        typeOfDisplay: "recentMovies",
        url: `${API_URL}/${API_VERSION}/discover/movie?api_key=${API_KEY}&language=fr-Fr&sort_by=release_date.desc&release_date.lte=${currentDate}&page=1`
      }
    }

    if (value === "popularMovies") {
      return {
        ...state,
        typeOfDisplay: "popularMovies",
        url: `${API_URL}/${API_VERSION}/movie/popular?api_key=${API_KEY}&language=fr-FR&page=1`
      }
    }

    if (value === "favorites") {
      return {
        ...state,
        typeOfDisplay: "favorites"
      }
    }
  }

  return state
}
