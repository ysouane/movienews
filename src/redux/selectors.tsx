export const checkTypeOfDisplay = (state: any) => state.typeOfDisplay
export const checkUrl = (state: any) => state.url
