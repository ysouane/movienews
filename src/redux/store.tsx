import { createStore } from "redux"
import { API_KEY, API_URL, API_VERSION } from "helpers"
import currentDate from "utils/currentDateUTCFormatted"
import Reducer from "./reducer"

export const initialState: any = {
  typeOfDisplay: "recentMovies",
  url: `${API_URL}/${API_VERSION}/discover/movie?api_key=${API_KEY}&language=fr-Fr&sort_by=release_date.desc&release_date.lte=${currentDate}&page=1`,
  favorites: []
}

export const store = createStore(Reducer, initialState)
