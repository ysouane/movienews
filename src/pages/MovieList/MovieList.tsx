import React, { useState, useEffect, useRef } from "react"
import { useSelector, useDispatch } from "react-redux"
import { changeTypeOfDisplayAndUrl } from "redux/actions"
import { checkTypeOfDisplay, checkUrl } from "redux/selectors"
import { API_KEY, API_URL, API_VERSION } from "helpers"
import currentDate from "utils/currentDateUTCFormatted"
import Card from "components/Card"
import SelectDisplay from "components/SelectDisplay"
import "./MovieList.scss"

const TYPE_OF_DISPLAY = {
  RECENT_MOVIES: "recentMovies",
  POPULAR: "popularMovies",
  FAVORITES: "favorites"
}

const MovieList = (props: any) => {
  const { reachedBottom } = props
  const dispatch = useDispatch()
  const typeOfDisplay = useSelector(checkTypeOfDisplay)
  const url = useSelector(checkUrl)
  const page = useRef(1)
  const [data, setData] = useState<any>([])
  const [isLoading, setLoading] = useState(false)
  const [favorites, setFavorites] = useState<any>([])

  async function fetchData(url: string) {
    setLoading(true)

    await fetch(url)
      .then((res) => res.json())
      .then((resJson) => setData(resJson))
      .catch((error) => {
        throw error
      })
      .finally(() => setLoading(false))
  }

  useEffect(() => {
    setData(fetchData(url))
  }, [url])

  const ChangeDisplay = (e: any) => {
    const { value } = e.target
    dispatch(changeTypeOfDisplayAndUrl(value))
  }

  async function fetchMoreData() {
    let url = ""

    if (typeOfDisplay === TYPE_OF_DISPLAY.RECENT_MOVIES) {
      url = `${API_URL}/${API_VERSION}/discover/movie?api_key=${API_KEY}&language=fr-FR&sort_by=release_date.desc&release_date.lte=${currentDate}&page=${page.current}`
    }

    if (typeOfDisplay === TYPE_OF_DISPLAY.POPULAR) {
      url = `${API_URL}/${API_VERSION}/movie/popular?api_key=${API_KEY}&language=fr-FR&page=${page.current}`
    }

    setLoading(true)

    await fetch(url)
      .then((res) => res.json())
      .then((resJson) =>
        setData({ results: [...data.results, ...resJson.results] })
      )
      .catch((error) => {
        throw error
      })
      .finally(() => setLoading(false))
  }

  useEffect(() => {
    if (reachedBottom === true && typeOfDisplay !== TYPE_OF_DISPLAY.FAVORITES) {
      page.current += 1
      fetchMoreData()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [reachedBottom])

  async function fetchFavorites(url: string) {
    await fetch(url)
      .then((res) => res.json())
      .then((resJson: any) => {
        setFavorites((favorites: never) => [...favorites, resJson])
      })
      .catch((error) => {
        throw error
      })
  }

  useEffect(() => {
    if (typeOfDisplay === TYPE_OF_DISPLAY.FAVORITES) {
      setFavorites([])
      let favorites = window.localStorage.movies

      if (favorites !== undefined) {
        favorites = window.localStorage.movies.split(",")
        for (let i = 0; i < favorites.length; i += 1) {
          fetchFavorites(
            `${API_URL}/${API_VERSION}/movie/${favorites[i]}?api_key=${API_KEY}&language=fr-FR`
          )
        }
      }
    }
  }, [typeOfDisplay])

  if (data?.results?.length > 1 || favorites?.length > 1) {
    return (
      <div className="movieList">
        <div className="movieList__selectDisplay">
          <SelectDisplay
            changeDisplay={ChangeDisplay}
            selected={typeOfDisplay}
          />
        </div>

        <div className="movieList__card">
          {typeOfDisplay === TYPE_OF_DISPLAY.RECENT_MOVIES ||
          typeOfDisplay === TYPE_OF_DISPLAY.POPULAR
            ? data?.results?.map((res: any) => {
                return (
                  <div key={res.id}>
                    <Card
                      posterPath={res.poster_path}
                      backdropPath={res.backdrop_path}
                      title={res.title}
                      release_date={res.release_date}
                      movieId={res.id}
                    />
                  </div>
                )
              })
            : favorites.map((res: any) => {
                return (
                  <div key={res.id}>
                    <Card
                      posterPath={res.poster_path}
                      backdropPath={res.backdrop_path}
                      title={res.title}
                      release_date={res.release_date}
                      movieId={res.id}
                    />
                  </div>
                )
              })}
        </div>
      </div>
    )
  }

  return (
    <Card
      posterPath={favorites?.poster_path}
      backdropPath={favorites?.backdrop_path}
      title={favorites?.title}
      release_date={favorites?.release_date}
      movieId={favorites?.id}
    />
  )
}

export default MovieList
