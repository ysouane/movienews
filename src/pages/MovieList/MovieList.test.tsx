import React from "react"
import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { store } from "redux/store"
import { BrowserRouter } from "react-router-dom"
import renderer from "react-test-renderer"
import MovieList from "./MovieList"
import SelectDisplay from "components/SelectDisplay"

describe("MovieList page", () => {
  it("Should return without error", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <MovieList />
        </Provider>
      </BrowserRouter>
    )
  })
})

it("renders select component correctly", () => {
  const tree = renderer.create(<SelectDisplay />).toJSON()
  expect(tree).toMatchSnapshot()
})
