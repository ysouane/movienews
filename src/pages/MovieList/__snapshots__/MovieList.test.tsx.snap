// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders select component correctly 1`] = `
<label
  htmlFor="typeOfDisplay"
>
  <select
    className="selectDisplay"
    id="typeOfDisplay"
  >
    <option
      value="recentMovies"
    >
      Films r√©cent
    </option>
    <option
      value="popularMovies"
    >
      Films les plus populaires
    </option>
    <option
      value="favorites"
    >
      Favoris
    </option>
  </select>
</label>
`;
